firestore-bigquery-export/scripts/import/src/run.ts

```diff
@@ -0,0 +1,167 @@
+import { cpus } from "os";
+import * as firebase from "firebase-admin";
+import { pool } from "workerpool";
+
+import { CliConfig } from "./types";
+import { parseConfig } from "./config";
+
+import {
+  ChangeType,
+  FirestoreBigQueryEventHistoryTracker,
+  FirestoreDocumentChangeEvent,
+} from "@firebaseextensions/firestore-bigquery-change-tracker";
+
+/**
+ * Import data from a collection group in parrallel using workers.
+ */
+async function processCollectionGroup(config: CliConfig): Promise<number> {
+  const maxWorkers = Math.ceil(cpus().length / 2);
+  const workerPool = pool(__dirname + "/worker.js", {
+    maxWorkers,
+    forkOpts: {
+      env: {
+        PROJECT_ID: config.projectId,
+        GOOGLE_CLOUD_PROJECT: config.projectId,
+        GCLOUD_PROJECT: config.projectId,
+        ...process.env,
+      },
+    },
+  });
+
+  const query = firebase
+    .firestore()
+    .collectionGroup(config.sourceCollectionPath);
+
+  const partitionsList = query.getPartitions(config.batchSize);
+
+  let total = 0;
+  let partitions = 0;
+
+  while (true) {
+    const inProgressTasks =
+      workerPool.stats().activeTasks + workerPool.stats().pendingTasks;
+    if (inProgressTasks >= maxWorkers) {
+      await new Promise((resolve) => setTimeout(resolve, 150));
+      continue;
+    }
+
+    // @ts-ignore, iterator not typed correctly.
+    const { value: partition, done } = await partitionsList.next();
+    if (done || !partition) {
+      break;
+    }
+
+    partitions++;
+
+    const query = partition.toQuery();
+
+    const serializedQuery = {
+      startAt: query._queryOptions.startAt,
+      endAt: query._queryOptions.endAt,
+      limit: query._queryOptions.limit,
+      offset: query._queryOptions.offset,
+    };
+
+    workerPool
+      .exec("processDocuments", [serializedQuery, config])
+      .then((count) => {
+        total += count;
+        console.log(`${total} documents processed`);
+      })
+      .catch((error) => {
+        // TODO handle, what should happen on error?
+        console.error(error);
+        process.exit(1);
+      });
+  }
+
+  // Wait for all tasks to be complete.
+  while (workerPool.stats().activeTasks + workerPool.stats().pendingTasks > 0) {
+    await new Promise((resolve) => setTimeout(resolve, 150));
+  }
+
+  await workerPool.terminate();
+
+  console.log(`Imported ${total} documents in ${partitions} partitions.`);
```

Comment:
Consider adding the bigquery dataset name to the output, and maybe even a link to the Cloud console:

```js
`https://console.cloud.google.com/bigquery?project=${config.projectId}`
```

Reply:
Link added